test/e2e/tests/eg_upgrade.go

```diff
@@ -267,46 +277,60 @@ func updateChartCRDs(actionConfig *action.Configuration, gatewayChart *chart.Cha
 	return err
 }
 
-// TODO: proper migration framework required
-func migrateChartCRDs(actionConfig *action.Configuration, gatewayChart *chart.Chart, timeout time.Duration) error {
+func migrateChartCRDs(actionConfig *action.Configuration, gatewayChart *chart.Chart, _ time.Duration) error {
 	crds, err := extractCRDs(actionConfig, gatewayChart)
 	if err != nil {
 		return err
 	}
 
+	// https: //gateway-api.sigs.k8s.io/guides/?h=upgrade#v12-upgrade-notes
+	storedVersionsMap := map[string]string{
+		"referencegrants.gateway.networking.k8s.io": "v1beta1",
+		"grpcroutes.gateway.networking.k8s.io":      "v1",
+	}
+
+	restCfg, err := actionConfig.RESTClientGetter.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	cli, err := client.New(restCfg, client.Options{})
+	if err != nil {
+		return err
+	}
+
 	for _, crd := range crds {
-		if crd.Name == "backendtlspolicies.gateway.networking.k8s.io" ||
-			crd.Name == "grpcroutes.gateway.networking.k8s.io" {
-			newVersion, err := getGWAPIVersion(crd.Object)
+		storedVersion, ok := storedVersionsMap[crd.Name]
+		if !ok {
+			continue
+		}
+
+		newVersion, err := getGWAPIVersion(crd.Object)
+		if err != nil {
+			return err
+		}
+
+		if strings.HasPrefix(newVersion, "v1.2.0") {
+			existingCRD := &apiextensionsv1.CustomResourceDefinition{}
+			err := cli.Get(context.Background(), types.NamespacedName{Name: crd.Name}, existingCRD)
+			if kerrors.IsNotFound(err) {
+				continue
+			}
+			if err != nil {
+				return fmt.Errorf("failed to get CRD: %s", err.Error())
+			}
+
+			// previous version exists
+			existingVersion, err := getGWAPIVersion(existingCRD)
 			if err != nil {
 				return err
 			}
-			// https://gateway-api.sigs.k8s.io/guides/?h=upgrade#v11-upgrade-notes
-			if newVersion == "v1.2.0-rc2" {
-				helper := resource.NewHelper(crd.Client, crd.Mapping)
-				existingCRD, err := helper.Get(crd.Namespace, crd.Name)
-				if kerrors.IsNotFound(err) {
-					continue
-				}
 
-				// previous version exists
-				existingVersion, err := getGWAPIVersion(existingCRD)
-				if err != nil {
-					return err
-				}
+			if existingVersion == "v1.0.0" {
+				existingCRD.Status.StoredVersions = []string{storedVersion}
```

Comment:
Right now, gw-api recommends to delete CRDs to completely: https://gateway-api.sigs.k8s.io/guides/?h=upgrade#v11-upgrade-notes

Should we change our instructions to users to reflect the upgrade path that is tested by EG?  https://github.com/envoyproxy/gateway/blob/0f4cb27f0110051f7811122a2dd73652c93a15b2/site/content/en/v1.1/install/install-yaml.md?plain=1#L41

Reply:
here should be `1.1` -> `1.2`, let me try to remove it.